hoomd/mpcd/CollisionMethod.cc

```diff
@@ -48,15 +48,284 @@ void mpcd::CollisionMethod::collide(uint64_t timestep)
+            {
+            continue;
+            }
+        // get velocities and masses
+        const Scalar4 vel_mass_const = h_velocity.data[particle_index];
+        const vec3<Scalar> vel_const(vel_mass_const);
+        const Scalar mass_const = vel_mass_const.w;
+
+        // get displacement
+        const Scalar4 postype_const = h_postype.data[particle_index];
+        const vec3<Scalar> pos_const(postype_const);
+        const Scalar4 postype_central = h_postype.data[central_idx];
+        const vec3<Scalar> pos_central(postype_central);
+        vec3<Scalar> displacement = pos_const - pos_central;
+
+        // change in linear and angular momentum
+        const vec3<Scalar> initial_vel_const(h_initial_vel.data[idx]);
+        const vec3<Scalar> linmom_change = mass_const * (vel_const - initial_vel_const);
+        const vec3<Scalar> angmom_change = cross(displacement, linmom_change);
+
+        // accumulate onto central particle
+        h_linmom_accum.data[central_idx] += vec_to_scalar3(linmom_change);
+        h_angmom_accum.data[central_idx] += vec_to_scalar3(angmom_change);
+        }
     }
 
+void mpcd::CollisionMethod::transferRigidBodyCollision(uint64_t timestep)
+    {
+    ArrayHandle<Scalar3> h_linmom_accum(m_linmom_accum, access_location::host, access_mode::read);
+    ArrayHandle<Scalar3> h_angmom_accum(m_angmom_accum, access_location::host, access_mode::read);
+    ArrayHandle<Scalar4> h_velocity(m_pdata->getVelocities(),
+                                    access_location::host,
+                                    access_mode::readwrite);
+    ArrayHandle<Scalar4> h_angmom(m_pdata->getAngularMomentumArray(),
+                                  access_location::host,
+                                  access_mode::readwrite);
+    ArrayHandle<Scalar4> h_orientation(m_pdata->getOrientationArray(),
+                                       access_location::host,
+                                       access_mode::read);
+    ArrayHandle<unsigned int> h_body(m_pdata->getBodies(),
+                                     access_location::host,
+                                     access_mode::read);
+    ArrayHandle<unsigned int> h_rtag(m_pdata->getRTags(), access_location::host, access_mode::read);
+    // add accumulated momentum to the central particle
+    unsigned int num_total = m_pdata->getN();
+    for (unsigned int idx = 0; idx < num_total; ++idx)
+        {
+        // check that the particle is in a rigid body and a central particle
+        const unsigned int central_tag = h_body.data[idx];
+        if (central_tag >= MIN_FLOPPY)
+            {
+            continue;
+            }
+        const unsigned int central_idx = h_rtag.data[central_tag];
+        if (central_idx != idx)
+            {
+            continue;
+            }
+        // get accumulated momentum for particle
+        vec3<Scalar> linmom_accum(h_linmom_accum.data[idx]);
+        vec3<Scalar> angmom_accum(h_angmom_accum.data[idx]);
```

Comment:
The linear momentum can stay as a `Scalar3`.
```suggestion

        // get accumulated momentum for particle
        const Scalar3 linmom_accum(h_linmom_accum.data[idx]);
        const vec3<Scalar> angmom_accum(h_angmom_accum.data[idx]);
```